### ISSUE TITLE
The order of the effects changed in SolidJS 1.5

### ISSUE THREAD

**Issue by juanrgm** [solidjs/solid#1199]
### Describe the bug

Hi all. 

I am facing a problem after upgrading to SolidJS 1.5 (https://github.com/swordev/suid/issues/92) and I don't know if this behavior change is documented.

### Your Example Website or App

https://playground.solidjs.com/?version=1.5.3#NobwRAdghgtgpmAXGGUCWEwBowBcCeADgsrgM4Ae2YZA9gK4BOAxiWGjIbY7gAQi9GcCABM4jXgF9eAM0a0YvADo1aAGzQiAtACsyAegDucAEYqA3EogcuPfr2ZCouOAFEZMuM1xYHTlwDKaADm0Gq+tBAAsgwQfNJyCsqqGtp6FlZWcBS2fGIyUPRqfDL0EN5okbwAwrEujAAUAJT8VrwOkWR8wDB1cCK+ZHC4MWUuIgC6vAC8fnDOcEGhUGoNBWpDTZYQbbyRo3ENzTMAfK0Q7e1DI30iDbiM9HBbu5IvOxdzC+6e3kct0zOIF27WYnXUcAAdGpaMEGipemN+ipfIi4v1mu92m9trshLgmBcADwiNAANxOIDR42akNwtACDwwcKakiJ+lJFO2kkyECEonE-1OvCJtSREn0J18IlozHo8DikOCw1cajgCtwACF8ABJO4qKCEQgqJoAQneYEkEyAA

## Code

```ts
import { render } from "solid-js/web";
import { createEffect, createSignal, onMount } from "solid-js";

export default function Counter() {
  const [mounted, setMounted] = createSignal(false);

  onMount(() => {
    setMounted(true);
  });

  createEffect(() => {
    console.log("mounted", mounted());
  });

  return <div>{mounted().toString()}</div>;
}

render(() => <Counter />, document.getElementById("app")!);
```

## SolidJS: 1.5.3

https://playground.solidjs.com/?version=1.5.3#NobwRAdghgtgpmAXGGUCWEwBowBcCeADgsrgM4Ae2YZA9gK4BOAxiWGjIbY7gAQi9GcCABM4jXgF9eAM0a0YvADo1aAGzQiAtACsyAegDucAEYqA3EogcuPfr2ZCouOAFEZMuM1xYHTlwDKaADm0Gq+tBAAsgwQfNJyCsqqGtp6FlZWcBS2fGIyUPRqfDL0EN5okbwAwrEujAAUAJT8VrwOkWR8wDB1cCK+ZHC4MWUuIgC6vAC8fnDOcEGhUGoNBWpDTZYQbbyRo3ENzTMAfK0Q7e1DI30iDbiM9HBbu5IvOxdzC+6e3kct0zOIF27WYnXUcAAdGpaMEGipemN+ipfIi4v1mu92m9trshLgmBcADwiNAANxOIDR42akNwtACDwwcKakiJ+lJFO2kkyECEonE-1OvCJtSREn0J18IlozHo8DikOCw1cajgCtwACF8ABJO4qKCEQgqJoAQneYEkEyAA

> mounted true

## SolidJS 1.4.8

https://playground.solidjs.com/?version=1.4.8#NobwRAdghgtgpmAXGGUCWEwBowBcCeADgsrgM4Ae2YZA9gK4BOAxiWGjIbY7gAQi9GcCABM4jXgF9eAM0a0YvADo1aAGzQiAtACsyAegDucAEYqA3EogcuPfr2ZCouOAFEZMuM1xYHTlwDKaADm0Gq+tBAAsgwQfNJyCsqqGtp6FlZWcBS2fGIyUPRqfDL0EN5okbwAwrEujAAUAJT8VrwOkWR8wDB1cCK+ZHC4MWUuIgC6vAC8fnDOcEGhUGoNBWpDTZYQbbyRo3ENzTMAfK0Q7e1DI30iDbiM9HBbu5IvOxdzC+6e3kct0zOIF27WYnXUcAAdGpaMEGipemN+ipfIi4v1mu92m9trshLgmBcADwiNAANxOIDR42akNwtACDwwcKakiJ+lJFO2kkyECEonE-1OvCJtSREn0J18IlozHo8DikOCw1cajgCtwACF8ABJO4qKCEQgqJoAQneYEkEyAA

> mounted false
> mounted true

### SolidJS 1.5.3

https://playground.solidjs.com/?version=1.5.3#NobwRAdghgtgpmAXGGUCWEwBowBcCeADgsrgM4Ae2YZA9gK4BOAxiWGjIbY7gAQi9GcCABM4jLLwDKAC1oB3XgF9eAM0a0YvADo1aAGzQiAtACsyAenlwARroDc2iBy49+vZkKi44AUVWqcMy4kp5w3nBSaADm0PqStBAAsgwQfCrqmjp6hibmDk5OcBSufGKqUPT6fKr0EMFoibwAwqk+jAAUAJT8TrweiWR8wDBtcCKSZHC4KXU+IgC6vAC8Hl4+UbFQ+h0V+lNdjhD9+tOCcKoAatv0cIi8UBD4K7wQVfqFx7yJs2kd3SsAHy9L79KYzMYiDq4Ri3Q59ZTwiAIsIRfyBYL-HrLYEgBH9ZiDAxwAB0+lo0Q6ulGc3GukkNLS426kiEVxucCR-SUSIRQlwTGOVNBvAAPLIFLx5DJhMsQIz5t0lID8f0xSI0AA3c6qOVs676W7K+WQ7ok3C0KQwjCUrpKUUWDWalUih0S+Qu-pIpSfISicRYoFi1q0xi8CyAyQiWjMejwNIk6LTXyneO4ABC+AAklDdFBCIRdF0AIRIsBKBZAA

### SolidJS 1.4.8

https://playground.solidjs.com/?version=1.4.8#NobwRAdghgtgpmAXGGUCWEwBowBcCeADgsrgM4Ae2YZA9gK4BOAxiWGjIbY7gAQi9GcCABM4jLLwDKAC1oB3XgF9eAM0a0YvADo1aAGzQiAtACsyAenlwARroDc2iBy49+vZkKi44AUVWqcMy4kp5w3nBSaADm0PqStBAAsgwQfCrqmjp6hibmDk5OcBSufGKqUPT6fKr0EMFoibwAwqk+jAAUAJT8TrweiWR8wDBtcCKSZHC4KXU+IgC6vAC8Hl4+UbFQ+h0V+lNdjhD9+tOCcKoAatv0cIi8UBD4K7wQVfqFx7yJs2kd3SsAHy9L79KYzMYiDq4Ri3Q59ZTwiAIsIRfyBYL-HrLYEgBH9ZiDAxwAB0+lo0Q6ulGc3GukkNLS426kiEVxucCR-SUSIRQlwTGOVNBvAAPLIFLx5DJhMsQIz5t0lID8f0xSI0AA3c6qOVs676W7K+WQ7ok3C0KQwjCUrpKUUWDWalUih0S+Qu-pIpSfISicRYoFi1q0xi8CyAyQiWjMejwNIk6LTXyneO4ABC+AAklDdFBCIRdF0AIRIsBKBZAA

### Expected behavior

> mounted false
> mounted true

### Platform

- OS: Windows
- Browser: Chrome
- Version: 104

**Comment by ryansolid** [comment]
Yes this was the batching change we advertised in the release. https://github.com/solidjs/solid/releases/tag/v1.5.0

Which is the big change here. It's not that the effects run in a different order, but that changes are applied immediately.  In general using effects to synchronize state isn't great and there is a tradeoff here. The new behavior removes something that kept on tripping up developers as Solid in some places would update signals immediately and others not. And it wasn't something they opted into. 

I honestly think an argument for the old behavior from a design perspective makes a lot of sense and I made that argument for years. But it did really complicate everything from a learning perspective as well. 

The simple solution is to queue your setter on the next microtask to get close to the previous behavior back.

**Comment by yw662** [comment]
So does it mean `{ defer: true }` is no longer needed ?

**Comment by juanrgm** [comment]
I understand the background but, for me and now (no offense), SolidJS does magic. I will need to relearn SolidJS.

**Comment by yw662** [comment]
Is it like if createEffect is before onMount, the result is different?

If so it could be a problem I guess.

**Comment by ryansolid** [comment]
> Is it like if createEffect is before onMount, the result is different?
> 
> If so it could be a problem I guess.

Yeah this is the one part if there was one that I feel the least happy about. However it does solve a bunch of other issues. 

> I understand the background but, for me and now (no offense), SolidJS does magic. I will need to relearn SolidJS.

@juanrgm I can respect that. Because it does change the expectations. Interestingly enough this was always under the hood. The changes to batch were actually just deleting a bunch of stuff. Trying to force the old behavior was a bit unnatural but something I thought necessary. But it was constantly a source of issues. This wasn't an easy decision and we spent several months debating this in the discord. Probably should have engaged the RFC process earlier (like we are with Solid 2.0). But this seems to be the behavior that aligns the most with other people's expectations. I say debate but it was more like myself alone defending the old behavior and everyone else wanting this. After exploring many different options, writing an article about it: https://dev.to/this-is-learning/the-cost-of-consistency-in-ui-frameworks-4agi and testing a different alternative we didn't ultimately adopt here we are.

We need more docs I admit but the funny thing about it was for the most part we only needed to do deletion in the docs as well and new Solid developers won't be tripped up. This was a huge simplification. But I empathize with those comfortable with the previous behavior.

**Comment by fweth** [comment]
Just FWIW I wanted to mention that component level `onCleanup` is always called last (after the effect cleanups), but `onCleanup` inside `onMount` is called in the order of when `onMount` was called.  So I think there is some logic to it, but one could maybe reflect in the documentation that `onMount` is just a short form of `createEffect` with `untrack` (so no special priority) but component level `onCleanup` is always guaranteed to run last.

**Comment by fweth** [comment]
FWIW I just wanted to mention that component level `onCleanup` is always called last.  So `onMount` is just a short form for `createEffect` with `untrack`, you can have [multiple `onMounts`](https://playground.solidjs.com/?hash=1544681271&version=1.5.3) being called before, after or between the effects and their cleanup functions will execute in the same order.  I think there is some good logic behind it, but maybe the docs could reflect on that a little bit better, especially that component level `onCleanup` is not the perfectly symmetrical counterpart to `onMount`.

**Comment by modderme123** [comment]
One thing that we recommend everyone do is avoid setting state from most computations/hooks/effects/memos and ideally restructure your code to derive the state with a memo or at least make it explicit in your code that you expect the setter to happen next frame with requestAnimationFrame or requestIdeCallback. This may not be possible in all cases, and if you can't seem to figure out how to restructure your code, I'd love to see a minimal example to figure out what useful primitives solid core can provide or to help restructure your code.

**Comment by juanrgm** [comment]
The original code was written in ReactJS and it was ported automatically to SolidJS with `@suid/codemod`.

Prior to SolidJS 1.5 the behavior was compatible with ReactJS, now I don't know if I can automate it.

I don't know if there are other similar cases derived from this change.

**Comment by yw662** [comment]
Maybe find all onMount and wrap the inner function with raf/timeout.

**Comment by ryansolid** [comment]
It comes down to the setSignal calls from `onMount` or `createEffect`. onMount it is pretty safe to wrap the set calls in a `queueMicrotask` but effects are a bit trickier perhaps.
